server/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractPercentilesAggregationBuilder.java

```diff
@@ -0,0 +1,361 @@
+import org.elasticsearch.search.aggregations.support.CoreValuesSourceType;
+import org.elasticsearch.search.aggregations.support.ValueType;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceAggregationBuilder;
+import org.elasticsearch.search.aggregations.support.ValuesSourceParserHelper;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * This provides a base class for aggregations that are building percentiles or percentiles-like functionality (e.g. percentile ranks).
+ * It provides a set of common fields/functionality for setting the available algorithms (TDigest and HDRHistogram),
+ * as well as algorithm-specific settings via a {@link PercentilesConfig} object
+ */
+public abstract class AbstractPercentilesAggregationBuilder<T extends AbstractPercentilesAggregationBuilder<T>>
+    extends ValuesSourceAggregationBuilder.LeafOnly<ValuesSource, T> {
+
+    public static final ParseField KEYED_FIELD = new ParseField("keyed");
+    protected boolean keyed = true;
+    protected double[] values;
+    private PercentilesConfig percentilesConfig;
+    private ParseField valuesField;
+
+    public static <T extends AbstractPercentilesAggregationBuilder<T>> ConstructingObjectParser<T, String> getParser(String aggName,
+                                                         TriFunction<String, double[], PercentilesConfig, T> fn,
+                                                         ParseField valuesField) {
+
+        /**
+         * This is a non-ideal ConstructingObjectParser, because it is a compromise between Percentiles and Ranks.
+         * Ranks requires an array of values because there is no sane default, and we want to keep that in the ctor.
+         * Percentiles has defaults, which means the API allows the user to either use the default or configure
+         * their own.
+         *
+         * The mutability of Percentiles keeps us from having a strict ConstructingObjectParser, while the ctor
+         * of Ranks keeps us from using a regular ObjectParser.
+         *
+         * This is a compromise, in that it is a ConstructingOP which accepts all optional arguments, and then we sort
+         * out the behavior from there
+         *
+         * `args` are provided from the ConstructingObjectParser in-order they are defined in the parser.  So:
+         *  - args[0]: values
+         *  - args[1]: tdigest config options
+         *  - args[2]: hdr config options
+         *
+         *  If `args` is null or empty, it means all were omitted.  This is usually an anti-pattern for
+         *  ConstructingObjectParser, but we're allowing it because of the above-mentioned reasons
+         */
+        ConstructingObjectParser<T, String> parser = new ConstructingObjectParser<>(aggName, false, (args, name) -> {
```

Comment:
Could you pass a method that builds the default instead of having it be handled in the ctor?